### PR TITLE
Add support for get style definition

### PR DESCRIPTION
**Pull request by cnmlgbgithub**
# PR Details

Create a new GetStyle function to get style definitions

## Description

Support to get the fills, theme color, font family, font size, alignment, cell protections, build-in, and custom number formats, implement this feature in the styles.go source code

## Related Issue

#314, #1520 and #1521

## How Has This Been Tested

All existing unit tests passed and new unit test cases passed.

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: styles.go [hunk 1]
```diff
@@ -1032,6 +1032,65 @@ func (f *File) NewStyle(style *Style) (int, error) {
 	return setCellXfs(s, fontID, numFmtID, fillID, borderID, applyAlignment, applyProtection, alignment, protection)
 }
 
+// Value extracts string data type text from a attribute value.
```
Comment (xuri) [inline styles.go:1035]: I suggest that move these types of functions implemented to the `lib.go` source file.

File: styles.go [hunk 2]
```diff
@@ -1032,6 +1032,65 @@ func (f *File) NewStyle(style *Style) (int, error) {
...
+			if fnt.Name != nil {
+				font.Family = fnt.Name.Value()
+			}
+			if fnt.Sz != nil {
+				font.Size = fnt.Sz.Value()
+			}
+			if fnt.Strike != nil {
+				font.Strike = fnt.Strike.Value()
+			}
+			font.Color = strings.TrimPrefix(fnt.Color.RGB, "FF")
```
Comment (xuri) [inline styles.go:1085]: We need to check and make sure the `font.Color` was not a null pointer before accessing the following 5 color-related fields.

**Review by xuri — CHANGES_REQUESTED** [review]

File: styles.go (hunk 1 above)
Comment (cnmlgbgithub) [reply styles.go]: Okay. I have moved these to the `lib.go` source file.

File: styles.go (hunk 2 above)
Comment (cnmlgbgithub) [reply styles.go]: Done.

**Review by enzshew — APPROVED** [review]

File: styles.go [hunk 3]
```diff
@@ -1032,6 +1032,103 @@ func (f *File) NewStyle(style *Style) (int, error) {
...
+		xf.BorderID != nil && s.Borders != nil &&
+		*xf.BorderID < len(s.Borders.Border) {
+		if bdr := s.Borders.Border[*xf.BorderID]; bdr != nil {
+			var borders []Border
+			extractBorder := func(lineType string, line xlsxLine) {
+				if line.Style != "" {
+					borders = append(borders, Border{
+						Type:  lineType,
+						Color: f.getThemeColor(line.Color),
+					})
```
[the remark this answers is not in the document]
Comment (cnmlgbgithub) [reply styles.go:1154]: Ok, and I have added support for getting fills style formats.

**Review by xuri — CHANGES_REQUESTED** [review]
LGTM. We need to implement `extractNumFmt`, `extractAlignment` and `extractProtection` in this change list.

File: styles.go [hunk 4]
```diff
@@ -1032,6 +1032,276 @@ func (f *File) NewStyle(style *Style) (int, error) {
...
+				}
+			}
+		}
+	}
+}
+
+// extractAlignment provides a function to extract alignment format by
+// given style definition.
+func (f *File) extractAlignment(xf xlsxXf, s *xlsxStyleSheet, style *Style) {
+	style.Alignment = &Alignment{
```
Comment (xuri) [inline styles.go:1280]: We need check and make sure the `ApplyAlignment` field was not nail before visit value of it.

File: styles.go [hunk 5]
```diff
@@ -1032,6 +1032,276 @@ func (f *File) NewStyle(style *Style) (int, error) {
...
+		TextRotation:    xf.Alignment.TextRotation,
+		Vertical:        xf.Alignment.Vertical,
+		WrapText:        xf.Alignment.WrapText,
+	}
+}
+
+// extractProtection provides a function to extract protection settings by
+// given format definition.
+func (f *File) extractProtection(xf xlsxXf, s *xlsxStyleSheet, style *Style) {
+	style.Protection = &Protection{}
```
Comment (xuri) [inline styles.go:1296]: Same as above, checking `ApplyProtection` if nil.

File: styles_test.go [hunk 6]
```diff
@@ -488,3 +488,105 @@ func TestGetNumFmtID(t *testing.T) {
...
+	style, err = f.GetStyle(styleID)
+	assert.NoError(t, err)
+	assert.Equal(t, expected.Fill, style.Fill)
+
+	expected = &Style{NumFmt: 27}
+	styleID, err = f.NewStyle(expected)
+	assert.NoError(t, err)
+	style, err = f.GetStyle(styleID)
+	assert.NoError(t, err)
+	assert.Equal(t, expected.NumFmt, style.NumFmt)
```
Comment (xuri) [inline styles_test.go:558]: This unit test was failed. I suggest create a new function `isLangNumFmt` to check if a given number format ID is a built-in language glyphs number format code, note that the ID range is 27-36, 50-62, 67-81.

File: styles.go [hunk 7]
```diff
@@ -1032,6 +1032,307 @@ func (f *File) NewStyle(style *Style) (int, error) {
...
+			style.Font = &font
+		}
+	}
+}
+
+// extractNumFmt provides a function to extract number format by given styles
+// definition.
+func (f *File) extractNumFmt(xf xlsxXf, s *xlsxStyleSheet, style *Style) {
+	if xf.NumFmtID != nil {
+		if _, ok := builtInNumFmt[*xf.NumFmtID]; ok {
```
Comment (xuri) [inline styles.go:1249]: Checking number format ID with `isLangNumFmt` at here.

File: styles.go [hunk 8]
```diff
@@ -1032,6 +1032,307 @@ func (f *File) NewStyle(style *Style) (int, error) {
...
+			style.NumFmt = *xf.NumFmtID
+			return
+		}
+		if s.NumFmts != nil {
+			for _, numFmt := range s.NumFmts.NumFmt {
+				for _, langFmtCodes := range langNumFmt {
+					if fmtCode, ok := langFmtCodes[*xf.NumFmtID]; ok && fmtCode == numFmt.FormatCode {
```
Comment (xuri) [inline styles.go:1256]: This statements can be removed.

**Review by xuri — CHANGES_REQUESTED** [review]
Good job. I've left some comments.

**Review by xuri — APPROVED** [review]
LGTM, thanks for your contribution.

**Pull request merged**